Source/Core/Core/HW/DSP.cpp

```diff
@@ -472,129 +489,137 @@ void UpdateAudioDMA()
+{
+  address &= 0x3ffffe0;
+  if (address >= 32 * 1024 * 1024)
+  {
+    return std::nullopt;
+  }
+  return (address & 0xff800000) >> 1 | (address & 0x003fffff);
+}
+
+using ARAM_ADDRESS_CONVERSION_F = std::optional<u32> (*)(u32 address);
+static ARAM_ADDRESS_CONVERSION_F conversion_functions[8] = {
+    ARAM_02MB_to_16MB, ARAM_04MB_to_16MB, ARAM_08MB_to_16MB, ARAM_16MB_to_16MB,
+    ARAM_32MB_to_16MB, ARAM_32MB_to_16MB, ARAM_32MB_to_16MB, ARAM_32MB_to_16MB,
+};
+
+constexpr u32 ARAM_LINE_SIZE = 0x20;
+constexpr u32 ARAM_MAX_TRANSFER_CHUNKING = 0x10 * ARAM_LINE_SIZE;  // 0x10 lines
 static void Do_ARAM_DMA()
 {
   s_dspState.DMAState = 1;
 
+  DEBUG_LOG(DSPINTERFACE, "DMA %08x bytes %s ARAM %08x to MRAM %08x PC: %08x",
+            s_arDMA.Cnt.dir ? "from" : "to", s_arDMA.Cnt.count, s_arDMA.ARAddr, s_arDMA.MMAddr, PC);
+  // ARAM is mirrored every 64MB (verified on real HW) - done in address conversion func
+  // Source/destination/count aligned to 32 bytes      - done in MMIO handler
+
+  u32 bytes_to_transfer = std::min(s_arDMA.Cnt.count, ARAM_MAX_TRANSFER_CHUNKING);
   // ARAM DMA transfer rate has been measured on real hw
-  int ticksToTransfer = (s_arDMA.Cnt.count / 32) * 246;
-  CoreTiming::ScheduleEvent(ticksToTransfer, s_et_CompleteARAM);
+  int ticksToTransfer = (bytes_to_transfer / 32) * 246;
+  CoreTiming::ScheduleEvent(ticksToTransfer, s_et_ContinueARAM);
 
-  // Real hardware DMAs in 32byte chunks, but we can get by with 8byte chunks
   if (s_arDMA.Cnt.dir)
   {
     // ARAM -> MRAM
-    DEBUG_LOG(DSPINTERFACE, "DMA %08x bytes from ARAM %08x to MRAM %08x PC: %08x",
-              s_arDMA.Cnt.count, s_arDMA.ARAddr, s_arDMA.MMAddr, PC);
-
-    // Outgoing data from ARAM is mirrored every 64MB (verified on real HW)
-    s_arDMA.ARAddr &= 0x3ffffff;
-    s_arDMA.MMAddr &= 0x3ffffff;
-
-    if (s_arDMA.ARAddr < s_ARAM.size)
+    for (u32 n = 0; n < bytes_to_transfer; n += ARAM_LINE_SIZE)
     {
-      while (s_arDMA.Cnt.count)
+      std::optional<u32> physical_aram_addr =
+          conversion_functions[s_ARAM_Info.base_size](s_arDMA.ARAddr);
+      if (physical_aram_addr)
       {
-        // These are logically separated in code to show that a memory map has been set up
-        // See below in the write section for more information
-        if ((s_ARAM_Info.Hex & 0xf) == 3)
-        {
-          Memory::Write_U64_Swap(*(u64*)&s_ARAM.ptr[s_arDMA.ARAddr & s_ARAM.mask], s_arDMA.MMAddr);
-        }
-        else if ((s_ARAM_Info.Hex & 0xf) == 4)
-        {
-          Memory::Write_U64_Swap(*(u64*)&s_ARAM.ptr[s_arDMA.ARAddr & s_ARAM.mask], s_arDMA.MMAddr);
-        }
-        else
-        {
-          Memory::Write_U64_Swap(*(u64*)&s_ARAM.ptr[s_arDMA.ARAddr & s_ARAM.mask], s_arDMA.MMAddr);
-        }
-
-        s_arDMA.MMAddr += 8;
-        s_arDMA.ARAddr += 8;
-        s_arDMA.Cnt.count -= 8;
+        std::memcpy(Memory::GetPointer(s_arDMA.MMAddr), &s_ARAM.ptr[*physical_aram_addr],
+                    ARAM_LINE_SIZE);
```

Comment:
It seems the outer `if`/`else` branches each have the same `for` loop that only differs in the first line of the inner `if`/`else` branches.
Would it be worthwhile to pull the loop up (writing it just once) and pushing the outer `if`/`else` inside (which would become a "pick source/dest" for the `physical_aram_addr` case; and a conditional "do/don't" otherwise)?